repair/row_level.cc

```diff
@@ -591,50 +592,67 @@ static void add_to_repair_meta_for_followers(repair_meta& rm);
-                    row_list.push_back(repair_row({}, {}, dk_ptr, {}, is_dirty_on_master::no, std::move(mf)));
+            auto& mutation_fragments = x.get_mutation_fragments();
+            if (is_master) {
+                for (auto fmfit = mutation_fragments.begin(); fmfit != mutation_fragments.end(); fmfit = mutation_fragments.erase(fmfit)) {
+                    auto fmf = std::move(*fmfit);
+
+                    _metrics.rx_row_nr += 1;
+                    _metrics.rx_row_bytes += fmf.representation().size();
+                    // Keep the mutation_fragment in repair_row as an
+                    // optimization to avoid unfreeze again when
+                    // mutation_fragment is needed by _repair_writer.do_write()
+                    // to apply the repair_row to disk
+                    auto mf = make_lw_shared<mutation_fragment>(fmf.unfreeze(*s, permit));
+                    auto hash = hasher.do_hash_for_mf(*dk_ptr, *mf);
+                    position_in_partition pos(mf->position());
+                    row_list.push_back(repair_row(std::move(fmf), std::move(pos), dk_ptr, std::move(hash), is_dirty_on_master::yes, std::move(mf)));
+                    co_await coroutine::maybe_yield();
+                }
+            } else {
+                last_mf = {};
+                for (auto fmfit = mutation_fragments.begin(); fmfit != mutation_fragments.end(); fmfit = mutation_fragments.erase(fmfit)) {
+                    auto fmf = std::move(*fmfit);
+
+                    _metrics.rx_row_nr += 1;
+                    _metrics.rx_row_bytes += fmf.representation().size();
+                    auto mf = make_lw_shared<mutation_fragment>(fmf.unfreeze(*s, permit));
+                    // If the mutation_fragment has the same position as
+                    // the last mutation_fragment, it means they are the
+                    // same row with different contents. We can not feed
+                    // such rows into the sstable writer. Instead we apply
+                    // the mutation_fragment into the previous one.
+                    if (last_mf && cmp(last_mf->position(), mf->position()) == 0 && last_mf->mergeable_with(*mf)) {
+                        last_mf->apply(*s, std::move(*mf));
+                    } else {
+                        last_mf = mf;
+                        // On repair follower node, only decorated_key_with_hash and the mutation_fragment inside repair_row are used.
+                        row_list.push_back(repair_row({}, {}, dk_ptr, {}, is_dirty_on_master::no, std::move(mf)));
+                    }
+                    co_await coroutine::maybe_yield();
                 }
-                co_await coroutine::maybe_yield();
             }
+            co_await coroutine::maybe_yield();
         }
-        co_await coroutine::maybe_yield();
+    } catch (...) {
+        ex = std::current_exception();
+    }
+    if (ex) {
```

Comment:
And here.  In most case, we won't catch the exception to clear something.

Reply:
I don't understand. The exception is caught a couple of rows above and is handled here (since we can't co_await in the catch clause, unfortunately)

Reply:
I meant comments on the trick why we are catching the exception to clear the rows. Because if we are not care about the latency caused by the destruction, there  is no need to do this at all.